src/common/utils.test.ts

```diff
@@ -250,3 +252,150 @@ describe("getFragmentsByType", () => {
+      NETWORK: "",
+      INFURA_API_KEY: "",
+      ALCHEMY_API_KEY: "",
+      ETHERSCAN_API_KEY: "",
+    };
+  });
+
+  it("should use the details provided as top priority", () => {
+    const options = {
+      network: "ropsten",
+      provider: "infura",
+      apiKey: "bb46da3f80e040e8ab73c0a9ff365d18",
+    } as ProviderDetails;
+    const provider = generateProvider(options) as any;
+
+    expect(provider?._network?.name).toEqual("ropsten");
+    expect(provider?.apiKey).toEqual("bb46da3f80e040e8ab73c0a9ff365d18");
+    expect(provider?.connection?.url).toMatch(/(infura)/i);
+  });
+
+  it("should use the default values to generate the provider if user did not specify any provider details", () => {
+    const provider = generateProvider() as any;
+    expect(provider?._network?.name).toEqual("homestead");
+    expect(provider?.apiKey).toEqual("bb46da3f80e040e8ab73c0a9ff365d18");
+    expect(provider?.connection?.url).toMatch(/(infura)/i);
+  });
+
+  it("should use the default etherscan apiKey if no apiKey specified", () => {
+    const options = {
+      network: "ropsten",
+      provider: "etherscan",
+    } as ProviderDetails;
+    const provider = generateProvider(options) as any;
+    expect(provider?.baseUrl).toMatch(/(etherscan)/i);
+    expect(provider?.apiKey).toEqual("3TDJ85CXAA4NCTBMP3UE98UR48NC2XKVGG");
+  });
+
+  it("should use the default alchemy apiKey if no apiKey specified", () => {
+    const options = {
+      network: "ropsten",
+      provider: "alchemy",
+    } as ProviderDetails;
+    const provider = generateProvider(options) as any;
+    expect(provider?.connection?.url).toMatch(/(alchemy)/i);
+    expect(provider?.apiKey).toEqual("OlOgD-8qs5l3pQm-B_fcrMAmHTmAwkGj");
+  });
+
+  it("should use the default jsonrpc url which is localhost:8545", () => {
+    const options = {
+      network: "ropsten",
+      provider: "jsonrpc",
+    } as ProviderDetails;
+    const provider = generateProvider(options) as any;
+    expect(provider?.connection?.url).toMatch(/(localhost:8545)/i);
+  });
+
+  it("should still generate a provider even if only one option (network) is provided", () => {
+    const options = { network: "ropsten" } as ProviderDetails;
+    const provider = generateProvider(options) as any;
+    expect(provider?._network?.name).toEqual("ropsten");
+    expect(provider?.apiKey).toEqual("bb46da3f80e040e8ab73c0a9ff365d18");
+    expect(provider?.connection?.url).toMatch(/(infura)/i);
+  });
+
+  it("should still generate a provider even if only one option (provider) is provided", () => {
+    const options = { provider: "infura" } as ProviderDetails;
+    const provider = generateProvider(options) as any;
+    expect(provider?._network?.name).toEqual("homestead");
+    expect(provider?.apiKey).toEqual("bb46da3f80e040e8ab73c0a9ff365d18");
+    expect(provider?.connection?.url).toMatch(/(infura)/i);
+  });
+
+  it("should still generate a provider even if only one option (url) is provided", () => {
```

Comment:
surprised this works, what does ethers.js do when you provide a url like that?

Reply:
actually it will just use this url, there wont be any error msg when generating the provider